versiondb/client/convert_to_sst.go

```diff
@@ -0,0 +1,280 @@
+			})
+
+			return err
+		}); err != nil {
+			break
+		}
+	}
+	close(inputChan)
+	if err != nil {
+		return err
+	}
+
+	if isEmpty {
+		// SSTFileWriter don't support writing empty files, so we stop early here.
+		return nil
+	}
+
+	sstWriter := newSSTFileWriter()
+	defer sstWriter.Destroy()
+
+	sstSeq := 0
+	openNextFile := func() error {
+		if err := sstWriter.Open(filepath.Join(sstDir, sstFileName(store, sstSeq))); err != nil {
+			return err
+		}
+		sstSeq++
+		return nil
+	}
+
+	if err := openNextFile(); err != nil {
+		return err
+	}
+
+	var lastKey []byte
+	for item := range outputChan {
+		ts, pair := decodeSorterItem(item)
+
+		// Only breakup sst file when the user-key(without timestamp) is different,
+		// because the rocksdb ingestion logic checks for overlap in keys without the timestamp part currently.
+		if sstWriter.FileSize() >= sstFileSize && !bytes.Equal(lastKey, pair.Key) {
+			if err := sstWriter.Finish(); err != nil {
+				return err
+			}
+			if err := openNextFile(); err != nil {
+				return err
+			}
+		}
+
+		key := cloneAppend(prefix, pair.Key)
+		if pair.Delete {
+			err = sstWriter.DeleteWithTS(key, ts)
+		} else {
+			err = sstWriter.PutWithTS(key, ts, pair.Value)
+		}
+		if err != nil {
+			return err
+		}
+
+		lastKey = pair.Key
+	}
+	return sstWriter.Finish()
+}
+
+// scanChangeSetFiles find change set files from the directory and sort them by the first version included, filter out
+// empty files.
+func scanChangeSetFiles(changeSetDir, store string) ([]FileWithVersion, error) {
+	// scan directory to find the change set files
+	storeDir := filepath.Join(changeSetDir, store)
+	entries, err := os.ReadDir(storeDir)
+	if err != nil {
+		return nil, err
+	}
+	fileNames := make([]string, len(entries))
+	for i, entry := range entries {
+		fileNames[i] = filepath.Join(storeDir, entry.Name())
+	}
+	return SortFilesByFirstVerson(fileNames)
+}
+
+// sstFileName inserts the seq integer into the base file name
+func sstFileName(store string, seq int) string {
+	return fmt.Sprintf("%s-%d%s", store, seq, SSTFileExtension)
+}
+
+func newSSTFileWriter() *grocksdb.SSTFileWriter {
+	envOpts := grocksdb.NewDefaultEnvOptions()
+	return grocksdb.NewSSTFileWriter(envOpts, tsrocksdb.NewVersionDBOpts(true))
+}
+
+// encodeSorterItem encode kv-pair for use in external sorter.
+//
+// layout: key + version(8) + delete(1) + [ value ] + key length(SizeKeyLength)
+// we put the key and version in the front of payload so it can take advantage of the delta encoding in the `ExtSorter`.
+func encodeSorterItem(version uint64, pair iavl.KVPair) []byte {
+	item := make([]byte, sizeOfSorterItem(pair))
+	copy(item, pair.Key)
+	offset := len(pair.Key)
+
+	binary.LittleEndian.PutUint64(item[offset:], version)
+	offset += tsrocksdb.TimestampSize
+
+	if pair.Delete {
+		item[offset] = 1
+		offset++
+	} else {
+		copy(item[offset+1:], pair.Value)
+		offset += len(pair.Value) + 1
+	}
+	binary.LittleEndian.PutUint32(item[offset:], uint32(len(pair.Key)))
```

Comment:
is this bound by uint32 -- I saw elsewhere sizes as uint64?

Reply:
max key/value length is uint32, both leveldb/rocksdb have that limit, it should be much smaller than that in practice, I'm proposing smaller limit at application level: https://github.com/cosmos/cosmos-sdk/pull/14645

Reply:
> I saw elsewhere sizes as uint64?

The size of an encoded change set of a whole version is encoded as uint64.